syntax = "proto3";

message pbft_msg
{
    pbft_msg_type type = 1;

    // These three will be present on every type of message other than requests, and identify the client request in question
    uint64 view = 2;
    uint64 sequence = 3;
    // TODO: Most messages should contain only the hash of the request - KEP-344
    pbft_request request = 4;

    // For prepare, commit, etc messages where the recipient needs to count the senders. This is the sender's uuid.
    string sender = 5;
}

enum pbft_msg_type {
    PBFT_MSG_TYPE_UNDEFINED = 0;
    PBFT_MSG_TYPE_REQUEST = 1;
    PBFT_MSG_TYPE_PREPREPARE = 2;
    PBFT_MSG_TYPE_PREPARE = 3;
    PBFT_MSG_TYPE_COMMIT = 4;
}

message pbft_request
{
    string operation = 1;
    uint64 timestamp = 2;
    string client = 3;
}
